File: assign1/src/rust/src/main.rs
use std::io::Write;
use std::time::Instant;

/* PAPI constants */
const PAPI_OK: i32 = 0;
const PAPI_L1_DCM: i32 = -2147483648;
const PAPI_L2_DCM: i32 = -2147483646;

fn main() {
    /* Init PAPI */
    let mut event_set: i32 = papi_sys::PAPI_NULL;
    let mut values: [i64; 2] = [0, 0];
    unsafe {
        let r = papi_sys::PAPI_library_init(papi_sys::PAPI_VER_CURRENT);
        if r != papi_sys::PAPI_VER_CURRENT {
            println!("PAPI library version mismatch!");
            return;
        }
        papi_sys::PAPI_create_eventset(&mut event_set);
        papi_sys::PAPI_add_event(event_set, PAPI_L1_DCM);
        papi_sys::PAPI_add_event(event_set, PAPI_L2_DCM);
    }

    /* Operations */
    let mut operation: u8 = 0;
    loop {
        println!("");
        println!("0. Exit");
        println!("1. Multiplication");
        println!("2. Line Multiplication");
        println!("3. Block Multiplication");
        print!("Selection?: ");
        std::io::stdout().flush().unwrap();
        let mut str_op_buf = String::new();
        std::io::stdin()
            .read_line(&mut str_op_buf)
            .expect("Failed to read from stdin");
        let op_buf = str_op_buf.trim();
        match op_buf.parse::<u8>() {
            Ok(i) => operation = i,
            Err(..) => println!("Invalid unsigned integer"),
        }
        if operation == 0 {
            break;
        }
        let mut m_size: usize = 0;
        let mut str_m_size_buf = String::new();
        print!("Dimensions: matrixSizes=cols ? ");
        std::io::stdout().flush().unwrap();
        std::io::stdin()
            .read_line(&mut str_m_size_buf)
            .expect("Failed to read from stdin");
        let m_buf = str_m_size_buf.trim();
        match m_buf.parse::<usize>() {
            Ok(i) => m_size = i,
            Err(..) => println!("Invalid matrix size"),
        };

        unsafe {
            papi_sys::PAPI_start(event_set);
        }

        match operation {
            1 => mult(m_size),
            2 => mult_line(m_size),
            3 => mult_line(m_size),
            _ => println!("Invalid option!"),
        }

        unsafe {
            if papi_sys::PAPI_stop(event_set, (&mut values) as *mut i64) != PAPI_OK {
                println!("ERROR: Stop PAPI");
            }
            println!("L1 DCM: {}", values[0]);
            println!("L2 DCM: {}", values[1]);
            if papi_sys::PAPI_reset(event_set) != PAPI_OK {
                println!("FAIL reset");
            }
        }
    }

    /* Destroy PAPI */
    unsafe {
        if papi_sys::PAPI_remove_event(event_set, PAPI_L1_DCM) != PAPI_OK {
            println!("FAIL remove event");
        }
        if papi_sys::PAPI_remove_event(event_set, PAPI_L2_DCM) != PAPI_OK {
            println!("FAIL remove event");
        }
        if papi_sys::PAPI_destroy_eventset(&mut event_set) != PAPI_OK {
            println!("FAIL destroy");
        }
    }
}

fn mult(m_size: usize) {
    let matrix_left: Vec<f64> = vec![1.0; m_size * m_size];
    let mut matrix_right: Vec<f64> = vec![0.0; m_size * m_size];
    let mut matrix_result: Vec<f64> = vec![0.0; m_size * m_size];

    for line in 0..m_size {
        for col in 0..m_size {
            matrix_right[line * m_size + col] = line as f64 + 1.0;
        }
    }

    let now = Instant::now();

    for line in 0..m_size {
        for col in 0..m_size {
            let mut temp: f64 = 0.0;
            for k in 0..m_size {
                temp += matrix_left[line * m_size + k] * matrix_right[k * m_size + col];
            }
            matrix_result[line * m_size + col] = temp;
        }
    }

    let elapsed = now.elapsed();
    println!("Time: {:3.3?}", elapsed);

    for col in 0..(std::cmp::min(10, m_size)) {
        print!("{} ", matrix_result[col])
    }
    println!("");
}

fn mult_line(m_size: usize) {
    let matrix_left: Vec<f64> = vec![1.0; m_size * m_size];
    let mut matrix_right: Vec<f64> = vec![0.0; m_size * m_size];
    let mut matrix_result: Vec<f64> = vec![0.0; m_size * m_size];

    for line in 0..m_size {
        for col in 0..m_size {
            matrix_right[line * m_size + col] = line as f64 + 1.0;
        }
    }

    let now = Instant::now();

    for line in 0..m_size {
        for iter in 0..m_size {
            for col in 0..m_size {
                matrix_result[line * m_size + col] +=
                    matrix_left[line * m_size + iter] * matrix_right[iter * m_size + col]
            }
        }
    }

    let elapsed = now.elapsed();
    println!("Time: {:3.3?}", elapsed);

    for col in 0..(std::cmp::min(10, m_size)) {
        print!("{} ", matrix_result[col])
    }
    println!("");
}

File: assign1/src/cpp/src/main.cpp
#include <stdio.h>
#include <iostream>
#include <iomanip>
#include <time.h>
#include <cstdlib>
#include <papi.h>
#include <string.h>

using namespace std;

#define SYSTEMTIME clock_t

void mult(int matrixSize)
{

    SYSTEMTIME time1, time2;

    char st[100];
    double temp;
    int i, j, k;

    double *leftMatrix, *rightMatrix, *resultMatrix;

    leftMatrix = (double *)malloc((matrixSize * matrixSize) * sizeof(double));
    rightMatrix = (double *)malloc((matrixSize * matrixSize) * sizeof(double));
    resultMatrix = (double *)malloc((matrixSize * matrixSize) * sizeof(double));

    for (i = 0; i < matrixSize; i++)
        for (j = 0; j < matrixSize; j++)
            leftMatrix[i * matrixSize + j] = (double)1.0;

    for (i = 0; i < matrixSize; i++)
        for (j = 0; j < matrixSize; j++)
            rightMatrix[i * matrixSize + j] = (double)(i + 1);

    time1 = clock();

    for (i = 0; i < matrixSize; i++)
    {
        for (j = 0; j < matrixSize; j++)
        {
            temp = 0;
            for (k = 0; k < matrixSize; k++)
            {
                temp += leftMatrix[i * matrixSize + k] * rightMatrix[k * matrixSize + j];
            }
            resultMatrix[i * matrixSize + j] = temp;
        }
    }

    time2 = clock();
    sprintf(st, "Time: %3.3f seconds\n", (double)(time2 - time1) / CLOCKS_PER_SEC);
    cout << st;

    // display 10 elements of the result matrix tto verify correctness
    cout << "Result matrix: " << endl;
    for (i = 0; i < 1; i++)
    {
        for (j = 0; j < min(10, matrixSize); j++)
            cout << resultMatrix[j] << " ";
    }
    cout << endl;

    free(leftMatrix);
    free(rightMatrix);
    free(resultMatrix);
}

// add code here for matrixSizee x matrixSizee matriz multiplication
void multLine(int matrixSize)
{

    SYSTEMTIME time1, time2;

    char st[100];
    double temp;
    int i, j, k;

    double *leftMatrix, *rightMatrix, *resultMatrix;

    leftMatrix = (double *)malloc((matrixSize * matrixSize) * sizeof(double));
    rightMatrix = (double *)malloc((matrixSize * matrixSize) * sizeof(double));
    resultMatrix = (double *)malloc((matrixSize * matrixSize) * sizeof(double));
    memset(resultMatrix, 0, (matrixSize * matrixSize) * sizeof(double));

    for (i = 0; i < matrixSize; i++)
        for (j = 0; j < matrixSize; j++)
            leftMatrix[i * matrixSize + j] = (double)1.0;

    for (i = 0; i < matrixSize; i++)
        for (j = 0; j < matrixSize; j++)
            rightMatrix[i * matrixSize + j] = (double)(i + 1);

    time1 = clock();

    for (i = 0; i < matrixSize; i++)
    {
        for (k = 0; k < matrixSize; k++)
        {
            for (j = 0; j < matrixSize; j++)
            {
                resultMatrix[i * matrixSize + j] += leftMatrix[i * matrixSize + k] * rightMatrix[k * matrixSize + j]; // leftMatrix[i][k] * rightMatrix[k][j]
            }
        }
    }

    time2 = clock();
    sprintf(st, "Time: %3.3f seconds\n", (double)(time2 - time1) / CLOCKS_PER_SEC);
    cout << st;

    // display 10 elements of the result matrix tto verify correctness
    cout << "Result matrix: " << endl;
    for (i = 0; i < 10; i++)
    {
        for (j = 0; j < min(10, matrixSize); j++)
            cout << resultMatrix[j] << " ";
    }
    cout << endl;

    free(leftMatrix);
    free(rightMatrix);
    free(resultMatrix);
}

// add code here for block x block matriz multiplication
void multBlock(int matrixSize, int bkSize) {}

void handle_error(int retval)
{
    printf("PAPI error %d: %s\n", retval, PAPI_strerror(retval));
    exit(1);
}

void init_papi()
{
    int retval = PAPI_library_init(PAPI_VER_CURRENT);
    if (retval != PAPI_VER_CURRENT && retval < 0)
    {
        printf("PAPI library version mismatch!\n");
        exit(1);
    }
    if (retval < 0)
        handle_error(retval);

    std::cout << "PAPI Version Number: MAJOR: " << PAPI_VERSION_MAJOR(retval)
              << " MINOR: " << PAPI_VERSION_MINOR(retval)
              << " REVISION: " << PAPI_VERSION_REVISION(retval) << "\n";
}

int main(int argc, char *argv[])
{

    char c;
    int matrixSize, blockSize;
    int op;

    int EventSet = PAPI_NULL;
    long long values[2];
    int ret;

    ret = PAPI_library_init(PAPI_VER_CURRENT);
    if (ret != PAPI_VER_CURRENT)
        std::cout << "FAIL" << endl;

    ret = PAPI_create_eventset(&EventSet);
    if (ret != PAPI_OK)
        cout << "ERROR: create eventset" << endl;

    ret = PAPI_add_event(EventSet, PAPI_L1_DCM);
    if (ret != PAPI_OK)
        cout << "ERROR: PAPI_L1_DCM" << endl;

    ret = PAPI_add_event(EventSet, PAPI_L2_DCM);
    if (ret != PAPI_OK)
        cout << "ERROR: PAPI_L2_DCM" << endl;

    op = 1;
    do
    {
        cout << endl
             << "0. Exit" << endl
             << "1. Multiplication" << endl;
        cout << "2. Line Multiplication" << endl;
        cout << "3. Block Multiplication" << endl;
        cout << "Selection?: ";
        cin >> op;
        if (op == 0)
            break;
        printf("Dimensions: matrixSizes=cols ? ");
        cin >> matrixSize;

        // Start counting
        ret = PAPI_start(EventSet);
        if (ret != PAPI_OK)
            cout << "ERROR: Start PAPI" << endl;

        switch (op)
        {
        case 1:
            mult(matrixSize);
            break;
        case 2:
            multLine(matrixSize);
            break;
        case 3:
            cout << "Block Size? ";
            cin >> blockSize;
            multBlock(matrixSize, blockSize);
            break;
        }

        ret = PAPI_stop(EventSet, values);
        if (ret != PAPI_OK)
            cout << "ERROR: Stop PAPI" << endl;
        printf("L1 DCM: %lld \n", values[0]);
        printf("L2 DCM: %lld \n", values[1]);

        ret = PAPI_reset(EventSet);
        if (ret != PAPI_OK)
            std::cout << "FAIL reset" << endl;

    } while (op != 0);

    ret = PAPI_remove_event(EventSet, PAPI_L1_DCM);
    if (ret != PAPI_OK)
        std::cout << "FAIL remove event" << endl;

    ret = PAPI_remove_event(EventSet, PAPI_L2_DCM);
    if (ret != PAPI_OK)
        std::cout << "FAIL remove event" << endl;

    ret = PAPI_destroy_eventset(&EventSet);
    if (ret != PAPI_OK)
        std::cout << "FAIL destroy" << endl;
}
